ostd/src/sync/spin.rs

```diff
@@ -16,90 +17,115 @@ use crate::{
 };
 
 /// A spin lock.
-pub struct SpinLock<T: ?Sized> {
+#[repr(transparent)]
+pub struct SpinLock<T: ?Sized, G = PreemptDisabled> {
+    phantom: PhantomData<G>,
+    /// Only the last field of a struct may have a dynamically sized type.
+    /// That's why SpinLockInner is put in the last field.
+    inner: SpinLockInner<T>,
+}
+
+struct SpinLockInner<T: ?Sized> {
     lock: AtomicBool,
     val: UnsafeCell<T>,
 }
 
-impl<T> SpinLock<T> {
+/// A guardian that denotes the guard behavior for holding the spin lock.
+pub trait Guardian {
+    type Guard;
+
+    /// Return a guard object
+    fn guard() -> Self::Guard;
+}
+
+/// A guardian that disables preemption while holding the spin lock.
+pub struct PreemptDisabled;
+
+impl Guardian for PreemptDisabled {
+    type Guard = DisablePreemptGuard;
+
+    fn guard() -> Self::Guard {
+        disable_preempt()
+    }
+}
+
+/// A guardian that disables IRQs while holding the spin lock.
+pub struct LocalIrqDisabled;
+
+impl Guardian for LocalIrqDisabled {
+    type Guard = DisabledLocalIrqGuard;
+
+    fn guard() -> Self::Guard {
+        disable_local()
+    }
+}
+
+impl<T, G: Guardian> SpinLock<T, G> {
     /// Creates a new spin lock.
     pub const fn new(val: T) -> Self {
-        Self {
+        let lock_inner = SpinLockInner {
             lock: AtomicBool::new(false),
             val: UnsafeCell::new(val),
+        };
+        Self {
+            phantom: PhantomData,
+            inner: lock_inner,
         }
     }
 }
 
-impl<T: ?Sized> SpinLock<T> {
-    /// Acquires the spin lock with disabling the local IRQs. This is the most secure
-    /// locking way.
+impl<T: ?Sized> SpinLock<T, PreemptDisabled> {
+    /// Converts the guard behavior from disabling preemption to disabling IRQs.
     ///
-    /// This method runs in a busy loop until the lock can be acquired.
-    /// After acquiring the spin lock, all interrupts are disabled.
-    pub fn lock_irq_disabled(&self) -> SpinLockGuard<T> {
-        let guard = disable_local();
-        self.acquire_lock();
-        SpinLockGuard_ {
-            lock: self,
-            inner_guard: InnerGuard::IrqGuard(guard),
-        }
-    }
-
-    /// Tries acquiring the spin lock immedidately with disabling the local IRQs.
-    pub fn try_lock_irq_disabled(&self) -> Option<SpinLockGuard<T>> {
-        let irq_guard = disable_local();
-        if self.try_acquire_lock() {
-            let lock_guard = SpinLockGuard_ {
-                lock: self,
-                inner_guard: InnerGuard::IrqGuard(irq_guard),
-            };
-            return Some(lock_guard);
-        }
-        None
+    /// This method would incur a certain time overhead over
+    /// [`lock`] itself. So prefer avoiding using this method when
+    /// IRQ handlers are allowed to get executed while holding this
+    /// lock. For example, if a lock is never used in the interrupt
+    /// context, then it is ok not to use this method in the process context.
```

Comment:
This paragraph seems more appropriate in the type documentation of `LocalIrqDisabled`?

The `disable_irq` method is one way to construct a `SpinLock<T, LocalIrqDisabled>`, but this method is not the only way to construct it (after all, it can be constructed directly with `SpinLock::new`). As long as the `SpinLock<T, LocalIrqDisabled>` is constructed, it will have the overhead over `SpinLock<T, PreemptionDisabled>`.

Just a reminder: Don't forget to update the wording after moving the paragraph, such as replacing `this method` with `this guardian` and replacing `` overhead over [`lock`] itself `` with `` overhead over [`PreemptionDisabled'] ``.